### PR TITLE
Serializing enums explicitly references types unless nullable (issue …

### LINKED ISSUE #1145: Enum serialization creates nullable type for non-nullable enums
Using an Enum as a key in a map yields a field with type <String?, dynamic> (rather than the expected <String, dynamic>.

```
import 'package:json_annotation/json_annotation.dart';

part 'shape_sorter.g.dart';

enum shapeProperty { size, color }

@JsonSerializable()
class ShapeSorter {
  Map<shapeProperty, String> shapeProperties;

  ShapeSorter(this.shapeProperties);
}
```
This leads to the generated code;
```
Map<String, dynamic> _$ShapeSorterToJson(ShapeSorter instance) =>
    <String, dynamic>{
      'shapeProperties': instance.shapeProperties
          .map((k, e) => MapEntry(_$shapePropertyEnumMap[k], e)),
    };
    

const _$shapePropertyEnumMap = {
  shapeProperty.size: 'size',
  shapeProperty.color: 'color',
};
```

This is because _$shapePropertyEnumMap[k] yields type String?, even though the value - I think - is guaranteed to exist in the map. This isn't a major problem unless you have post processing (I'm using FakeFirebaseFirestore, but there could be otherwise). 

This causes a problem with anything which is doing null-aware runtime typechecking for Map<String, dynamic>.

json_serializable: "6.2.0"

### PR DESCRIPTION
**Pull request by ssabdb**
#1145 I've fixed the issue for my own purposes by inspecting the `targetType` - if not nullable, I use the `!` operator to remove the nullable type outputted by directly indexing `_$EnumTypeEnumMap`.

This includes a unit test which fails prior to the fix.

I considered doing something similar to introducing a helper function used in the enum deserialization (`$enumDecode`), but this just seemed simpler - since the map  `_$EnumTypeEnumMap` is generated from the possible values of the enum, it should never result in a null, unless the variable storing the enum itself is nullable - and if that happens I don't emit an `!` operator.

It does change a number of other generated files though, which might be a concern. 

resolves #1145

### CODE REVIEW

**Comment by kevmoo** [comment]
@ssabdb – can you look at the failures?

**Comment by ssabdb** [comment]
Looks like the test builds some example files which are diff'd and have changed to include the null assertion “bang” operator as an intended result of this PR. 

They will need to be regenerated so the test passes - I'll figure it out today or tomorrow.

**Comment by ssabdb** [comment]
@kevmoo this failure looks related to the fix you're working on in #1117?

Let me know if I need to do anything else.

**Comment by kevmoo** [comment]
@ssabdb – the markdown failure is unrelated – I should likely get rid of that.

File: json_serializable/test/integration/integration_test.dart [hunk 1]
```diff
@@ -304,6 +304,14 @@ void main() {
     expect(dayTypeEnumValues, ['no-good', 'rotten', 'very-bad']);
   });
 
+  test('enum serialization has correct type', () {
+    // ignore: omit_local_variable_types
```
Comment (kevmoo) [inline json_serializable/test/integration/integration_test.dart:308]: Don't ignore this lint. Drop the local variable type!

File: json_serializable/test/integration/integration_test.dart [hunk 2]
```diff
@@ -304,6 +304,14 @@ void main() {
...
+    final Issue1145Regression cls =
+        Issue1145Regression(status: {Issue1145RegressionEnum.gamma: true});
+    // Not Map<String?, dynamic>
```
Comment (kevmoo) [inline json_serializable/test/integration/integration_test.dart:311]: Drop this comment. JSON does not allow null keys (at least I don't think so)

File: json_serializable/test/integration/integration_test.dart [hunk 3]
```diff
@@ -304,6 +304,14 @@ void main() {
...
+    expect(cls.toJson()['status'], const TypeMatcher<Map<String, dynamic>>());
```
Comment (kevmoo) [inline json_serializable/test/integration/integration_test.dart:312]: What is this before the change? I'm confused as to the actual new behavior...

**Review by kevmoo — CHANGES_REQUESTED** [review]

File: json_serializable/test/integration/integration_test.dart (hunk 3 above)
Comment (ssabdb) [reply json_serializable/test/integration/integration_test.dart:312]: The new behaviour is to return Map<String, dynamic> correctly, hence the typematcher. The previous behaviour was returning Map<String?, dynamic> - which as you say is not permitted by the json spec.

File: json_serializable/test/integration/integration_test.dart (hunk 2 above)
Comment (ssabdb) [reply json_serializable/test/integration/integration_test.dart:311]: I'm happy to clarify the comment, but checking that the resultant output is _not_ Map<String**?**, dynamic> is  the intent of the test

File: json_serializable/test/integration/integration_test.dart (hunk 3 above)
Comment (kevmoo) [reply json_serializable/test/integration/integration_test.dart:313]: ah! cool!

**Review by kevmoo — CHANGES_REQUESTED** [review]
Please add an entry to the changelog explaining this change, too!

**Comment by ssabdb** [comment]
@kevmoo. Done. I also added a second unit test to test the branching logic was working properly for nullable enums, eg. in a List type.

**Comment by kevmoo** [comment]
Still need a changelog entry then I'll run CI and land! Thanks so much!

**Comment by ssabdb** [comment]
No worries, I'd misunderstood changelog (I thought you were talking about the gitlog!)

I'm on the road at the moment until late next week but will have a shot at getting round to it.

File: json_serializable/CHANGELOG.md [hunk 4]
```diff
@@ -4,6 +4,7 @@
   of type `MyClass?`. ([#822](https://github.com/google/json_serializable.dart/issues/822))
 - Added support for `JsonSerializable(converters: <JsonConverter>[])`
   ([#1072](https://github.com/google/json_serializable.dart/issues/1072))
+- Fix issue with serialization of non-nullable enumerations to emit a non-nullable type ([#1146](https://github.com/google/json_serializable.dart/pull/1146))
```
Comment (kevmoo) [inline json_serializable/CHANGELOG.md:7]: Wrap this line at 80 columns!

**Review by kevmoo — DISMISSED** [review]

File: json_serializable/CHANGELOG.md (hunk 4 above)
Comment (ssabdb) [reply json_serializable/CHANGELOG.md]: Done!

**Pull request merged**